From 5a50e0a414b0932e7510a67f6ef39677982af5ad Mon Sep 17 00:00:00 2001
From: Flora Cui <Flora.Cui@amd.com>
Date: Thu, 12 Jan 2017 10:52:07 +0800
Subject: [PATCH 3073/3082] drm/amdgpu: refine amdgpu_sem implementation (v2)

don't alloc fd for each sem

v2: rebase on amd-staging (Alex)

Change-Id: I44e4f5e7e9793c6e7aeec3b27f89850587a39eda
Signed-off-by: Flora Cui <Flora.Cui@amd.com>
Acked-by: Hawking Zhang <Hawking.Zhang@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu.h     |   5 +
 drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c |   7 ++
 drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c | 191 ++++++++++++++++++++++----------
 drivers/gpu/drm/amd/amdgpu/amdgpu_sem.h |   8 +-
 include/uapi/drm/amdgpu_drm.h           |   4 +-
 5 files changed, 156 insertions(+), 59 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu.h b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
index 0b44d13..95c8417 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu.h
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu.h
@@ -57,6 +57,7 @@
 #include "amdgpu_sync.h"
 #include "amdgpu_ring.h"
 #include "amdgpu_vm.h"
+#include "amdgpu_sem.h"
 #include "amd_powerplay.h"
 #include "amdgpu_dpm.h"
 #include "amdgpu_acp.h"
@@ -828,6 +829,8 @@ struct amdgpu_fpriv {
 	struct mutex		bo_list_lock;
 	struct idr		bo_list_handles;
 	struct amdgpu_ctx_mgr	ctx_mgr;
+	spinlock_t		sem_handles_lock;
+	struct idr		sem_handles;
 };
 
 /*
@@ -1369,6 +1372,8 @@ int amdgpu_sem_ioctl(struct drm_device *dev, void *data,
 int amdgpu_sem_add_cs(struct amdgpu_ctx *ctx, struct amdgpu_ring *ring,
 		      struct amdgpu_sync *sync);
 
+void amdgpu_sem_destroy(struct amdgpu_fpriv *fpriv, u32 handle);
+
 /* VRAM scratch page for HDP bug, default vram page */
 struct amdgpu_vram_scratch {
 	struct amdgpu_bo		*robj;
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c
index 955384f..e44e360 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_kms.c
@@ -796,6 +796,8 @@ int amdgpu_driver_open_kms(struct drm_device *dev, struct drm_file *file_priv)
 
 	mutex_init(&fpriv->bo_list_lock);
 	idr_init(&fpriv->bo_list_handles);
+	spin_lock_init(&fpriv->sem_handles_lock);
+	idr_init(&fpriv->sem_handles);
 
 	amdgpu_ctx_mgr_init(&fpriv->ctx_mgr);
 
@@ -822,6 +824,7 @@ void amdgpu_driver_postclose_kms(struct drm_device *dev,
 	struct amdgpu_device *adev = dev->dev_private;
 	struct amdgpu_fpriv *fpriv = file_priv->driver_priv;
 	struct amdgpu_bo_list *list;
+	struct amdgpu_sem *sem;
 	int handle;
 
 	if (!fpriv)
@@ -852,6 +855,10 @@ void amdgpu_driver_postclose_kms(struct drm_device *dev,
 	idr_destroy(&fpriv->bo_list_handles);
 	mutex_destroy(&fpriv->bo_list_lock);
 
+	idr_for_each_entry(&fpriv->sem_handles, sem, handle)
+		amdgpu_sem_destroy(fpriv, handle);
+	idr_destroy(&fpriv->sem_handles);
+
 	kfree(fpriv);
 	file_priv->driver_priv = NULL;
 
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
index db16baa..aa33e8c 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
@@ -40,31 +40,17 @@ static int amdgpu_sem_cring_add(struct amdgpu_fpriv *fpriv,
 				struct drm_amdgpu_sem_in *in,
 				struct amdgpu_sem *sem);
 
-static const struct file_operations amdgpu_sem_fops;
-
-static struct amdgpu_sem *amdgpu_sem_alloc(struct fence *fence)
+static void amdgpu_sem_core_free(struct kref *kref)
 {
-	struct amdgpu_sem *sem;
-
-	sem = kzalloc(sizeof(struct amdgpu_sem), GFP_KERNEL);
-	if (!sem)
-		return NULL;
+	struct amdgpu_sem_core *core = container_of(
+		kref, struct amdgpu_sem_core, kref);
 
-	sem->file = anon_inode_getfile("sem_file",
-				       &amdgpu_sem_fops,
-				       sem, 0);
-	if (IS_ERR(sem->file))
-		goto err;
+	if (core->file)
+		fput(core->file);
 
-	kref_init(&sem->kref);
-	INIT_LIST_HEAD(&sem->list);
-	/* fence should be get before passing here */
-	sem->fence = fence;
-
-	return sem;
-err:
-	kfree(sem);
-	return NULL;
+	fence_put(core->fence);
+	mutex_destroy(&core->lock);
+	kfree(core);
 }
 
 static void amdgpu_sem_free(struct kref *kref)
@@ -72,15 +58,28 @@ static void amdgpu_sem_free(struct kref *kref)
 	struct amdgpu_sem *sem = container_of(
 		kref, struct amdgpu_sem, kref);
 
-	fence_put(sem->fence);
+	list_del(&sem->list);
+	kref_put(&sem->base->kref, amdgpu_sem_core_free);
 	kfree(sem);
 }
 
+static inline void amdgpu_sem_get(struct amdgpu_sem *sem)
+{
+	if (sem)
+		kref_get(&sem->kref);
+}
+
+static inline void amdgpu_sem_put(struct amdgpu_sem *sem)
+{
+	if (sem)
+		kref_put(&sem->kref, amdgpu_sem_free);
+}
+
 static int amdgpu_sem_release(struct inode *inode, struct file *file)
 {
 	struct amdgpu_sem *sem = file->private_data;
 
-	kref_put(&sem->kref, amdgpu_sem_free);
+	amdgpu_sem_put(sem);
 	return 0;
 }
 
@@ -102,49 +101,127 @@ static const struct file_operations amdgpu_sem_fops = {
 	.compat_ioctl = amdgpu_sem_file_ioctl,
 };
 
-static int amdgpu_sem_create(void)
+
+static inline struct amdgpu_sem *amdgpu_sem_lookup(struct amdgpu_fpriv *fpriv, u32 handle)
 {
-	return get_unused_fd_flags(O_CLOEXEC);
+	struct amdgpu_sem *sem;
+
+	spin_lock(&fpriv->sem_handles_lock);
+
+	/* Check if we currently have a reference on the object */
+	sem = idr_find(&fpriv->sem_handles, handle);
+	amdgpu_sem_get(sem);
+
+	spin_unlock(&fpriv->sem_handles_lock);
+
+	return sem;
+}
+
+static struct amdgpu_sem_core *amdgpu_sem_core_alloc(void)
+{
+	struct amdgpu_sem_core *core;
+
+	core = kzalloc(sizeof(*core), GFP_KERNEL);
+	if (!core)
+		return NULL;
+
+	kref_init(&core->kref);
+	mutex_init(&core->lock);
+	return core;
 }
 
-static int amdgpu_sem_signal(int fd, struct fence *fence)
+static struct amdgpu_sem *amdgpu_sem_alloc(void)
 {
 	struct amdgpu_sem *sem;
 
-	sem = amdgpu_sem_alloc(fence);
+	sem = kzalloc(sizeof(*sem), GFP_KERNEL);
 	if (!sem)
+		return NULL;
+
+	kref_init(&sem->kref);
+	INIT_LIST_HEAD(&sem->list);
+
+	return sem;
+}
+
+static int amdgpu_sem_create(struct amdgpu_fpriv *fpriv, u32 *handle)
+{
+	struct amdgpu_sem *sem;
+	struct amdgpu_sem_core *core;
+	int ret;
+
+	sem = amdgpu_sem_alloc();
+	core = amdgpu_sem_core_alloc();
+	if (!sem || !core) {
+		kfree(sem);
+		kfree(core);
 		return -ENOMEM;
-	fd_install(fd, sem->file);
+	}
+
+	sem->base = core;
+
+	idr_preload(GFP_KERNEL);
+	spin_lock(&fpriv->sem_handles_lock);
+
+	ret = idr_alloc(&fpriv->sem_handles, sem, 1, 0, GFP_NOWAIT);
 
+	spin_unlock(&fpriv->sem_handles_lock);
+	idr_preload_end();
+
+	if (ret < 0)
+		return ret;
+
+	*handle = ret;
 	return 0;
 }
 
-static int amdgpu_sem_wait(int fd, struct amdgpu_fpriv *fpriv,
+static int amdgpu_sem_signal(struct amdgpu_fpriv *fpriv,
+				u32 handle, struct fence *fence)
+{
+	struct amdgpu_sem *sem;
+	struct amdgpu_sem_core *core;
+
+	sem = amdgpu_sem_lookup(fpriv, handle);
+	if (!sem)
+		return -EINVAL;
+
+	core = sem->base;
+	mutex_lock(&core->lock);
+	fence_put(core->fence);
+	core->fence = fence_get(fence);
+	mutex_unlock(&core->lock);
+
+	amdgpu_sem_put(sem);
+	return 0;
+}
+
+static int amdgpu_sem_wait(struct amdgpu_fpriv *fpriv,
 			  struct drm_amdgpu_sem_in *in)
 {
-	struct file *file = fget(fd);
 	struct amdgpu_sem *sem;
-	int r;
+	int ret;
 
-	if (!file)
+	sem = amdgpu_sem_lookup(fpriv, in->handle);
+	if (!sem)
 		return -EINVAL;
 
-	sem = file->private_data;
-	if (!sem) {
-		r = -EINVAL;
-		goto err;
-	}
-	r = amdgpu_sem_cring_add(fpriv, in, sem);
-err:
-	fput(file);
-	return r;
+	ret = amdgpu_sem_cring_add(fpriv, in, sem);
+	amdgpu_sem_put(sem);
+
+	return ret;
 }
 
-static void amdgpu_sem_destroy(void)
+void amdgpu_sem_destroy(struct amdgpu_fpriv *fpriv, u32 handle)
 {
-	/* userspace should close fd when they try to destroy sem,
-	 * closing fd will free semaphore object.
-	 */
+	struct amdgpu_sem *sem = amdgpu_sem_lookup(fpriv, handle);
+	if (!sem)
+		return;
+
+	spin_lock(&fpriv->sem_handles_lock);
+	idr_remove(&fpriv->sem_handles, handle);
+	spin_unlock(&fpriv->sem_handles_lock);
+
+	kref_sub(&sem->kref, 2, amdgpu_sem_free);
 }
 
 static struct fence *amdgpu_sem_get_fence(struct amdgpu_fpriv *fpriv,
@@ -219,12 +296,14 @@ int amdgpu_sem_add_cs(struct amdgpu_ctx *ctx, struct amdgpu_ring *ring,
 	mutex_lock(&ctx->rings[ring->idx].sem_lock);
 	list_for_each_entry_safe(sem, tmp, &ctx->rings[ring->idx].sem_list,
 				 list) {
-		r = amdgpu_sync_fence(ctx->adev, sync, sem->fence);
-		fence_put(sem->fence);
+		r = amdgpu_sync_fence(ctx->adev, sync, sem->base->fence);
 		if (r)
 			goto err;
-		list_del(&sem->list);
-		kfree(sem);
+		mutex_lock(&sem->base->lock);
+		fence_put(sem->base->fence);
+		sem->base->fence = NULL;
+		mutex_unlock(&sem->base->lock);
+		list_del_init(&sem->list);
 	}
 err:
 	mutex_unlock(&ctx->rings[ring->idx].sem_lock);
@@ -238,14 +317,13 @@ int amdgpu_sem_ioctl(struct drm_device *dev, void *data,
 	struct amdgpu_fpriv *fpriv = filp->driver_priv;
 	struct fence *fence;
 	int r = 0;
-	int fd = args->in.fd;
 
 	switch (args->in.op) {
 	case AMDGPU_SEM_OP_CREATE_SEM:
-		args->out.fd = amdgpu_sem_create();
+		r = amdgpu_sem_create(fpriv, &args->out.handle);
 		break;
 	case AMDGPU_SEM_OP_WAIT_SEM:
-		r = amdgpu_sem_wait(fd, fpriv, &args->in);
+		r = amdgpu_sem_wait(fpriv, &args->in);
 		break;
 	case AMDGPU_SEM_OP_SIGNAL_SEM:
 		fence = amdgpu_sem_get_fence(fpriv, &args->in);
@@ -253,14 +331,15 @@ int amdgpu_sem_ioctl(struct drm_device *dev, void *data,
 			r = PTR_ERR(fence);
 			return r;
 		}
-		r = amdgpu_sem_signal(fd, fence);
+		r = amdgpu_sem_signal(fpriv, args->in.handle, fence);
 		fence_put(fence);
 		break;
 	case AMDGPU_SEM_OP_DESTROY_SEM:
-		amdgpu_sem_destroy();
+		amdgpu_sem_destroy(fpriv, args->in.handle);
 		break;
 	default:
-		return -EINVAL;
+		r = -EINVAL;
+		break;
 	}
 
 	return r;
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.h b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.h
index 56d59d3..04296ca 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.h
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.h
@@ -34,10 +34,16 @@
 #include <linux/spinlock.h>
 #include <linux/fence.h>
 
-struct amdgpu_sem {
+struct amdgpu_sem_core {
 	struct file		*file;
 	struct kref		kref;
 	struct fence            *fence;
+	struct mutex	lock;
+};
+
+struct amdgpu_sem {
+	struct amdgpu_sem_core	*base;
+	struct kref		kref;
 	struct list_head        list;
 };
 
diff --git a/include/uapi/drm/amdgpu_drm.h b/include/uapi/drm/amdgpu_drm.h
index 7542e89..d3e9040 100644
--- a/include/uapi/drm/amdgpu_drm.h
+++ b/include/uapi/drm/amdgpu_drm.h
@@ -227,7 +227,7 @@ union drm_amdgpu_vm {
 struct drm_amdgpu_sem_in {
 	/** AMDGPU_SEM_OP_* */
 	uint32_t        op;
-	int32_t         fd;
+	uint32_t	handle;
 	uint32_t        ctx_id;
 	uint32_t        ip_type;
 	uint32_t        ip_instance;
@@ -237,7 +237,7 @@ struct drm_amdgpu_sem_in {
 
 union drm_amdgpu_sem_out {
 	int32_t         fd;
-	uint32_t        _pad;
+	uint32_t        handle;
 };
 
 union drm_amdgpu_sem {
-- 
2.7.4

